File: 066_BestTimeToBuyAndSellStockWithTransactionFee.cpp
// Topic: Dynamic Programming / Greedy
// Best Time to Buy and Sell Stock with Transaction Fee: maintain hold and cash states
// Time Complexity: O(n)
// Space Complexity: O(1)

#include <bits/stdc++.h>
using namespace std;

int maxProfit(vector<int>& prices, int fee)
{
    long long hold = -prices[0], cash = 0;
    for (int i = 1; i < prices.size(); ++i)
    {
        cash = max(cash, hold + prices[i] - fee);
        hold = max(hold, cash - prices[i]);
    }
    return (int)cash;
}

int main()
{
    cout << maxProfit((vector<int>{1, 3, 2, 8, 4, 9}), 2) << "\n";   // 8
    cout << maxProfit((vector<int>{1, 3, 7, 5, 10, 3}), 3) << "\n";  // 6
    return 0;
}

File: 066_BestTimeToBuyAndSellStockWithTransactionFee.py
"""
Topic: Dynamic Programming / Greedy
Problem: Best Time to Buy and Sell Stock with Transaction Fee
Description: Track hold and cash states to maximize profit.
Time: O(n)
Space: O(1)
"""

from typing import List


def maxProfit(prices: List[int], fee: int) -> int:
    cash, hold = 0, -prices[0]
    for p in prices[1:]:
        cash = max(cash, hold + p - fee)
        hold = max(hold, cash - p)
    return cash


if __name__ == '__main__':
    print(maxProfit([1,3,2,8,4,9], 2))  # 8
    print(maxProfit([1,3,7,5,10,3], 3))  # 6
